pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+			return
+		}
+		job, ok = tombstone.Obj.(*batchv1.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ReplicaSet %#v", obj))
+			return
+		}
+	}
+
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef == nil {
+		// No controller should care about orphans being deleted.
+		return
+	}
+	cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *ControllerV2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *ControllerV2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.AddAfter(key, t)
+}
+
+// updateCronJob re-queues the CronJob for next scheduled time if there is a
+// change in spec.schedule otherwise it re-queues it now
+func (jm *ControllerV2) updateCronJob(old interface{}, curr interface{}) {
+	oldCJ, okOld := old.(*batchv1beta1.CronJob)
+	newCJ, okNew := curr.(*batchv1beta1.CronJob)
+
+	if !okOld || !okNew {
+		// typecasting of one failed, handle this better, may be log entry
+		return
+	}
+	// if the change in schedule results in next requeue having to be sooner than it already was,
+	// it will be handled here by the queue. If the next requeue is further than previous schedule,
+	// the sync loop will essentially be a no-op for the already queued key with old schedule.
+	if oldCJ.Spec.Schedule != newCJ.Spec.Schedule {
+		// schedule changed, change the requeue time
+		sched, err := cron.ParseStandard(newCJ.Spec.Schedule)
+		if err != nil {
+			// this is likely a user error in defining the spec value
+			// we should log the error and not reconcile this cronjob until an update to spec
+			klog.V(2).InfoS("unparseable schedule for cronjob", "cronjob", klog.KRef(newCJ.GetNamespace(), newCJ.GetName()), "schedule", newCJ.Spec.Schedule, "err", err)
+			jm.recorder.Eventf(newCJ, corev1.EventTypeWarning, "UnParseableCronJobSchedule", "unparseable schedule for cronjob: %s", newCJ.Spec.Schedule)
+			return
+		}
+		now := jm.now()
+		t := nextScheduledTimeDuration(sched, now)
+
+		jm.enqueueControllerWithTime(curr, *t)
+		return
+	}
+
+	// other parameters changed, requeue this now and if this gets triggered
+	// within deadline, sync loop will work on the CJ otherwise updates will be handled
+	// during the next schedule
+	// TODO: need to handle the change of spec.JobTemplate.metadata.labels explicitly
+	//   to cleanup jobs with old labels
+	jm.enqueueController(curr)
+}
+
+// TODO: @alpatel we need to return errors from this function, in order to allow
```

Comment:
this doesn't seem like a thing you can defer out of this PR.

Reply:
I think the function returns errors, I need to remove the TODO

Reply:
removed the TODO, thanks!